src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -271,6 +271,8 @@ public class HttpSource : IDisposable
             try
             {
                 response = await RetryHandler.SendAsync(request, log, cancellationToken);
+				
+                response.LogServerWarning(log);
```

Comment:
I think this should actually be done directly after the real network handler (`HttpClientHandler`) returns the actual response. The current approach has the following drawbacks:

1. If an HTTP 5XX is returned, the retry handler will send the request against before this line above can log.
1. If an HTTP 401 is returned, then the credential provider flow will occur before this warning can be logged.

Today the pipeline is something like this 

```
V3 resources -> HttpSource -> RetryHandler -> HttpClient -> custom HttpMessageHandler(s) -> HttpClientHandler
``` 

I think this logging should be done in an inner-most custom HttpMessageHandler:
```
V3 resources -> HttpSource -> RetryHandler -> HttpClient -> custom HttpHandler(s) -> warn -> HttpClientHandler
``` 

The `HttpMessageHandler` pipeline is defined here:
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpHandlerResourceV3Provider.cs